docs/man/mysqldbcompare.1
'\" t
.\"     Title: \fBmysqldbcompare\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 01/14/2017
.\"    Manual: MySQL Utilities
.\"    Source: MySQL 1.6.6
.\"  Language: English
.\"
.TH "\FBMYSQLDBCOMPARE\FR" "1" "01/14/2017" "MySQL 1\&.6\&.6" "MySQL Utilities"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
.HP \w'\fBmysqldbcompare\ \-\ 'u
mysqldbcompare \- Compare Two Databases and Identify Differences
.br
.\"[With ID Analytics extensions]
.SH "SYNOPSIS"
.HP \w'\fBmysqldbcompare\ 'u
\fBmysqldbcompare [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIdb1\fR\fR\fB [\fR\fB\fI:db2\fR\fR\fB] \&.\&.\&.\fR
.SH "DESCRIPTION"
.PP
This utility compares the objects and data from two databases to find differences\&. It identifies objects having different definitions in the two databases and presents them in a diff\-style format of choice\&. Differences in the data are shown using a similar diff\-style format\&. Changed or missing rows are shown in a standard format of GRID, CSV, TAB, or VERTICAL\&.
.PP
Use the notation db1:db2 to name two databases to compare, or, alternatively just db1 to compare two databases with the same name\&. The latter case is a convenience notation for comparing same\-named databases on different servers\&.
.PP
The comparison may be run against two databases of different names on a single server by specifying only the
\fB\-\-server1\fR
option\&. The user can also connect to another server by specifying the
\fB\-\-server2\fR
option\&. In this case, db1 is taken from server1 and db2 from server2\&.
.PP
All databases between two servers can also be compared using the
\fB\-\-all\fR
option\&. In this case, only the databases in common (with the same name) between the servers are successively compared\&. Therefore, no databases need to be specified but the
\fB\-\-server1\fR
and
\fB\-\-server2\fR
options are required\&. Users can skip the comparison of some of the databases using the
\fB\-\-exclude\fR
option\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
The data must not be changed during the comparison\&. Unexpected errors may occur if data is changed during the comparison\&.
.sp .5v
.RE
.PP
The objects considered in the database include tables, views, triggers, procedures, functions, and events\&. A count for each object type can be shown with the
\fB\-vv\fR
option\&.
.PP
The check is performed using a series of steps called tests\&. By default, the utility stops on the first failed test, but you can specify the
\fB\-\-run\-all\-tests\fR
option to cause the utility to run all tests regardless of their end state\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
Using
\fB\-\-run\-all\-tests\fR
may produce expected cascade failures\&. For example, if the row counts differ among two tables being compared, the data consistency also fails\&.
.sp .5v
.RE
.PP
The tests include the following:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Check database definitions
.sp
A database existence precondition check ensures that both databases exist\&. If they do not, no further processing is possible and the
\fB\-\-run\-all\-tests\fR
option is ignored\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Check existence of objects in both databases
.sp
The test for objects in both databases identifies those objects missing from one or another database\&. The remaining tests apply only to those objects that appear in both databases\&. To skip this test, use the
\fB\-\-skip\-object\-compare\fR
option\&. That can be useful when there are known missing objects among the databases\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  3." 4.2
.\}
Compare object definitions
.sp
The definitions (the
\fBCREATE\fR
statements) are compared and differences are presented\&. To skip this test, use the
\fB\-\-skip\-diff\fR
option\&. That can be useful when there are object name differences only that you want to ignore\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 4.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  4." 4.2
.\}
Check table row counts
.sp
This check ensures that both tables have the same number of rows\&. This does not ensure that the table data is consistent\&. It is merely a cursory check to indicate possible missing rows in one table or the other\&. The data consistency check identifies the missing rows\&. To skip this test, use the
\fB\-\-skip\-row\-count\fR
option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 5.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  5." 4.2
.\}
Check table data consistency
.sp
This check identifies both changed rows as well as missing rows from one or another of the tables in the databases\&. Changed rows are displayed as a diff\-style report with the format chosen (\fBGRID\fR
by default) and missing rows are also displayed using the format chosen\&. This check is divided in two steps: first the full table checksum is compared between the tables, then if this step fails (or is skipped) the algorithm to find rows differences is executed\&. To skip the preliminary checksum table step in this test, use the
\fB\-\-skip\-checksum\-table\fR
option\&. To skip this full test, use the
\fB\-\-skip\-data\-check\fR
option\&.
.RE
.PP
You may want to use the
\-\-skip\-xxx
options to run only one of the tests\&. This might be helpful when working to bring two databases into synchronization, to avoid running all of the tests repeatedly during the process\&.
.PP
Each test completes with one of the following states:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBpass\fR
.sp
The test succeeded\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBFAIL\fR
.sp
The test failed\&. Errors are displayed following the test state line\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBSKIP\fR
.sp
The test was skipped due to a missing prerequisite or a skip option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBWARN\fR
.sp
The test encountered an unusual but not fatal error\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\fR
.sp
The test is not applicable to this object\&.
.RE
.PP
To specify how to display diff\-style output, use one of the following values with the
\fB\-\-difftype\fR
option:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBunified\fR
(default)
.sp
Display unified format output\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBcontext\fR
.sp
Display context format output\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBdiffer\fR
.sp
Display differ\-style format output\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBsql\fR
.sp
Display SQL transformation statement output\&.
.RE
.PP
To specify how to display output for changed or missing rows, use one of the following values with the
\fB\-\-format\fR
option:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBgrid\fR
(default)
.sp
Display output in grid or table format like that of the
\fBmysql\fR
client command\-line tool\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBcsv\fR
.sp
Display output in comma\-separated values format\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBtab\fR
.sp
Display output in tab\-separated format\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBvertical\fR
.sp
Display output in single\-column format like that of the
\eG
command for the
\fBmysql\fR
client command\-line tool\&.
.RE
.PP
The
\fB\-\-changes\-for\fR
option controls the direction of the difference (by specifying the object to be transformed) in either the difference report (default) or the transformation report (designated with the
\fB\-\-difftype=sql\fR
option)\&. Consider the following command:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysqldbcompare \-\-server1=root@host1 \-\-server2=root@host2 \-\-difftype=sql db1:dbx\fR
.fi
.if n \{\
.RE
.\}
.PP
The leftmost database (db1) exists on the server designated by the
\fB\-\-server1\fR
option (host1)\&. The rightmost database (dbx) exists on the server designated by the
\fB\-\-server2\fR
option (host2)\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-changes\-for=server1\fR: Produce output that shows how to make the definitions of objects on
server1
like the definitions of the corresponding objects on
server2\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-changes\-for=server2\fR: Produce output that shows how to make the definitions of objects on
server2
like the definitions of the corresponding objects on
server1\&.
.RE
.PP
The default direction is
server1\&.
.PP
You must provide connection parameters (user, host, password, and so forth) for an account that has the appropriate privileges to access all objects in the operation\&.
.PP
If the utility is to be run on a server that has binary logging enabled, and you do not want the comparison steps logged, use the
\fB\-\-disable\-binary\-logging\fR
option\&.
.RE
.SS "OPTIONS"
.PP
\fBmysqldbcompare\fR
accepts the following command\-line options:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-all, \-a
.sp
Compare all database in common (with the same name) between two servers\&.
.sp
The
\fB\-\-all\fR
option ignores the following databases:
\fIINFORMATION_SCHEMA\fR,
\fIPERFORMANCE_SCHEMA\fR,
\fImysql\fR, and
\fIsys\fR\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
The
\fIsys\fR
database is ignored as of Utilities 1\&.6\&.2\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-help
.sp
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-license
.sp
Display license information and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-changes\-for=\fIdirection\fR
.sp
Specify the server to show transformations to match the other server\&. For example, to see the transformation for transforming object definitions on server1 to match the corresponding definitions on server2, use
\fB\-\-changes\-for=server1\fR\&. Permitted values are
\fBserver1\fR
and
\fBserver2\fR\&. The default is
\fBserver1\fR\&.
.RE
.sp
.RS 4
.ie n \{\

CHANGES.txt
Release-1.6.6-IDA
-------------
- COMT#c8dca61 : Add INFORMATION_SCHEMA-based event schema object handling

- COMT#befe5bd : ISSUE-36: Avoid using SHOW DATABASES for alive check
- COMT#7d4f198 : no mysql.event in mysql 8, remove it
- COMT#4e7bd0a : no more mysql.proc in mysql 8, use information_schema.routines
- COMT#43ac066 : use the pure python connector implementation
- COMT#34f707e : Include traceback output when error occurs
- COMT#20e6c0b : Properly quote the definer for CREATE / ALTER VIEW statements
- COMT#7a1c8b6 : sql_transform.py: fix skip-opt-autoinc
- COMT#513139a : mysqldiff: make the result sql more readable
- COMT#40e0b99 : mysqldiff: add --include-create & --include-drop options
- COMT#e74f19d : mysqldiff: add --skip-opt-autoinc
- COMT#65843c1 : sql_transform.py:_get_index_sql_clauses(): make indexes by name and index fields in consistent order


Release-1.6.5
-------------
- BUG#25030151 : mysqlfailover (v:1.6) was stopped silently when using a '--exec-before' script
- BUG#24742356 : typos in mysqlfailover messages
- BUG#24657793 : mysqldbcopy fails to copy views which refer other views
- BUG#24624716 : ISSUE WITH 'MYSQLREPLICATE' 1.6.4 UTILITY
- BUG#24613794 : MYSQLFRM UNABLE TO PARSE SOME .FRM FILES
- BUG#23598948 : tables with primary key of 0 with auto_increment fail to copy
- BUG#23578714 : fix the _conn_ipv4_num_only regular expression
- BUG#23335753 : mysqldbcopy issue with tinytext
- BUG#22738248 : mysqlfailover fails with master my.cnf[entry] style option and python 2.6
- BUG#22616847 : incorrect error message from mysqlrpladmin
- BUG#22543157 : mysqlbinlogpurge: some documentation requests
- BUG#22543116 : mysqlbinlogpurge requires the master port explicitly
- BUG#22457829 : mysql serverclone doesn't detect the error properly and fails with ambiguous msg
- BUG#22254445 : --verbosity on mysqfrm does't work
- BUG#22215626 : mysqlserverinfo requires connection option if show-servers used
- BUG#22155217 : mysqldbimport --skip does not work
- BUG#22093518 : mysqldbcopy fails to copy grants for test database
- BUG#22093453 : mysqldbcompare --quiet does not work
- BUG#21848362 : mysqlfailover ping option description
- BUG#21764003 : convenience notation not permitted
- BUG#21626041 : ls: cannot access /usr/share/pyshared/mysql: no such file or directory
- BUG#21497923 : mu1.5.5: message to be updated for invalid parameters to mysqlrplcheck.py
- BUG#21270853 : mysqlrplsync does not allow for a --exclude option
- BUG#21201454 : mysqlfailover do not set read_only to slave during preparing candidate promotion
- BUG#21068133 : add ssl connection section to clearly document ssl behavior
- BUG#20857151 : mysqldiff drops and recreates pk and indexes when it is not needed
- BUG#20807296 : remove use password() function (deprecated as of mysql 5.7.6)
- BUG#20524886 : mysqlfailover doesn't clean up failover_console table (ie:drop) in daemon mode
- BUG#20511130 : --force doesn't work for missing objects with mysqldiff
- BUG#20411237 : mysqlfrmreader must support new year data type
- BUG#20400663 : mysqldbcopy documented required permissions list is incomplete
- BUG#18763591 : using grid format header difference should be ignored for mysqldbcompare
- BUG#18755438 : pep8 issue on utilities code base
- BUG#18733683 : mysqldbexport table names in the output are not consistent
- BUG#18293794 : fix some errors on utilities code base and improve it

Release-1.6.4
-------------
- BUG#24348569 : utilities fail to properly detect 5.7.13 servers 
- BUG#23300260 : mysql utilities fail with "too many values to unpack" in _update_alias
- BUG#23111808 : mysqldbbcopy copy blob data when some field's value is null
- BUG#22932375 : ha utilties must set connection_timeout in c/py connection to short value
- BUG#22905132 : handling regexp for few of the utilities
- BUG#22878052 : add master outage delay for mysqlfailover
- BUG#22780854 : valueerror while using mysqlrplsync discover-slave option with host:port
- BUG#22764612 : valueerror: too many values to unpack " running mysqlrpladmin
- BUG#22619159 : special char in password causes error with mysqlrplshow
- BUG#22543517 : mysqlbinlogpurge does not work with other than the first replication channel
- BUG#22495193 : mysqldbexport does not dump bit fields correctly
- BUG#22393353 : mysql-utilities-1.5.6-osx10.9.dmg won't install under osx 10.11.2 (el capitan)
- BUG#22242601 : yum install of mysql-utilities-1.5.6-1.el6.src.rpm fails (documentation updated)
- BUG#22076859 : after pressing <delete> in mysql utility console the command buffer is corrupt
- BUG#21222808 : mysqldiff: encoding comments always make databases definitions look different
- BUG#17666275 : add authentication plugin info to mysqluserclone

Release-1.6.3
-------------
- BUG#22964855 : year needs updation in readme,license and mysqluc in mu 1.6.3 packages
- BUG#22781266 : exclude option in mysqldbexport does not seem to take complete effect
- BUG#22628187 : mysqldbcopy does not work with MySQL 5.7
- BUG#22617526 : mysqlfailover and mysqlrpladmin do not work with MySQL 5.7
- BUG#22608528 : mysqlrplsync does not work with MySQL 5.7
- BUG#21966831 : Typo in mysqlbinlogrotate documentation
- BUG#21905774 : missing man pages in rpm package 
- BUG#21845449 : Fix mysqldbcompare when comparing the same object on the same server
- BUG#21841353 : UnicodeEncodeError in mysqlauditgrep using unicode characters
- BUG#21839348 : AttributeError in mysqldbcompare using --all option
- BUG#21784538 : The --exclude option is not working for data in mysqldbexport
- BUG#21653496 : Daemon processes creates pidfiles with incorrect privileges
- BUG#21653352 : MySQL Utilities breaks when sql_mode="ANSI_QUOTES"
- BUG#21572065 : mysqldbcompare raise errors when the index definitions differs

scripts/mysqldbcompare.py
#!/usr/bin/env python
#
# Copyright (c) 2011, 2016, Oracle and/or its affiliates. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
#

"""
This file contains the operations to perform database consistency checking
on two databases.
"""

import os
import re
import sys

from mysql.utilities.common.tools import check_python_version
from mysql.utilities.exception import UtilError, FormatError
from mysql.utilities.command.dbcompare import (compare_all_databases,
                                               database_compare)
from mysql.utilities.common.ip_parser import parse_connection
from mysql.utilities.common.dbcompare import (DEFAULT_SPAN_KEY_SIZE,
                                              MAX_SPAN_KEY_SIZE)
from mysql.utilities.common.pattern_matching import (REGEXP_OBJ_NAME,
                                                     REGEXP_OBJ_NAME_AQ)
from mysql.utilities.common.tools import check_connector_python
from mysql.utilities.common.messages import (PARSE_ERR_DB_PAIR,
                                             PARSE_ERR_DB_PAIR_EXT,
                                             PARSE_ERR_DB_MISSING_CMP,
                                             PARSE_ERR_OPTS_REQ,
                                             PARSE_ERR_SPAN_KEY_SIZE_TOO_HIGH,
                                             PARSE_ERR_SPAN_KEY_SIZE_TOO_LOW,
                                             WARN_OPT_ONLY_USED_WITH)
from mysql.utilities.common.options import (add_difftype, add_regexp,
                                            add_verbosity, check_all,
                                            check_verbosity,
                                            add_changes_for, add_reverse,
                                            add_format_option,
                                            add_character_set_option,
                                            add_ssl_options, get_ssl_dict,
                                            setup_common_options,
                                            check_password_security,
                                            add_exclude, check_exclude_pattern)
from mysql.utilities.common.server import connect_servers
from mysql.utilities.common.sql_transform import (is_quoted_with_backticks,
                                                  remove_backtick_quoting,
                                                  quote_with_backticks)

# Check Python version compatibility
check_python_version()

# Constants
NAME = "MySQL Utilities - mysqldbcompare "
DESCRIPTION = "mysqldbcompare - compare databases for consistency"
USAGE = "%prog --server1=user:pass@host:port:socket " + \
        "--server2=user:pass@host:port:socket db1:db2"
PRINT_WIDTH = 75

# Check for connector/python
if not check_connector_python():
    sys.exit(1)

if __name__ == '__main__':
    # Setup the command parser
    parser = setup_common_options(os.path.basename(sys.argv[0]),
                                  DESCRIPTION, USAGE, server=False)

    # Connection information for the source server
    parser.add_option("--server1", action="store", dest="server1",
                      type="string", default=None,
                      help="connection information for first server in "
                           "the form: <user>[:<password>]@<host>[:<port>]"
                           "[:<socket>] or <login-path>[:<port>][:<socket>].")

    # Connection information for the destination server
    parser.add_option("--server2", action="store", dest="server2",
                      type="string", default=None,
                      help="connection information for second server in "
                           "the form: <user>[:<password>]@<host>[:<port>]"
                           "[:<socket>] or <login-path>[:<port>][:<socket>].")

    # Add character set option
    add_character_set_option(parser)

    # Output format
    add_format_option(parser, "display the output in either grid (default), "
                              "tab, csv, or vertical format", "grid")

    # Add skips
    parser.add_option("--skip-checksum-table", action="store_true",
                      dest="no_checksum_table",
                      help="skip CHECKSUM TABLE step in data consistency "
                           "check.")

    parser.add_option("--skip-object-compare", action="store_true",
                      dest="no_object_check",
                      help="skip object comparison step.")

    parser.add_option("--skip-row-count", action="store_true",
                      dest="no_row_count",
                      help="skip row count step.")

    parser.add_option("--skip-diff", action="store_true",
                      dest="no_diff",
                      help="skip the object diff step.")

    parser.add_option("--skip-data-check", action="store_true",
                      dest="no_data",
                      help="skip data consistency check.")

    # Skip check of table options.
    parser.add_option("--skip-table-options", action="store_true",
                      dest="skip_tbl_opts",
                      help="skip check of all table options (e.g., "
                           "AUTO_INCREMENT, ENGINE, CHARSET, etc.).")

    # Skip only AUTO_INCREMENT table option.
    parser.add_option("--skip-opt-autoinc", action="store_true", default=False,
                      dest="skip_opt_autoinc",
                      help="skip table option AUTO_INCREMENT only")

    # Skip table partitioning options
    parser.add_option("--skip-partitioning", action="store_true", default=False,
                      dest="skip_partitioning",
                      help="skip table partitioning options")

    # Add display width option
    parser.add_option("--width", action="store", dest="width",
                      type="int", help="display width",
                      default=PRINT_WIDTH)

    # run-all-tests mode
    parser.add_option("-t", "--run-all-tests", action="store_true",
                      dest="run_all_tests",
                      help="do not abort when a diff test fails")

    # Add the all database options
    parser.add_option("-a", "--all", action="store_true", dest="all",
                      default=False, help="check all databases")

    # Add the exclude database option
    add_exclude(parser, "databases", "db1", "db%")

    # Add compact option for resulting diff
    parser.add_option("-c", "--compact", action="store_true",
                      dest="compact", help="compact output from a diff.")

    # turn off binlog mode
    parser.add_option("--disable-binary-logging", action="store_true",
                      default=False, dest="toggle_binlog",
                      help="turn binary logging off during operation if "
                           "enabled (SQL_LOG_BIN=1). Note: may require SUPER "
                           "privilege. Prevents compare operations from being "
                           "written to the binary log.")

    # add the span key option
    parser.add_option(
        "--span-key-size", action="store", default=DEFAULT_SPAN_KEY_SIZE,
        type="int", dest="span_key_size",
        help="changes the size of the key used for compare table contents. A "
             "higher value can help to get more accurate results comparing "
             "large databases, but may slow the algorithm. Default value is "
             "{0}.".format(DEFAULT_SPAN_KEY_SIZE)
    )

    # add the use indexes option
    parser.add_option(
        "--use-indexes", action="store", type="string", default='',
        dest="use_indexes",
        help="for each table, indicate which index to use as if were a "
             "primary key (each of his columns must not allow null values)."
    )

    # Add verbosity and quiet (silent) mode
    add_verbosity(parser, True)

    # Add difftype option
    add_difftype(parser, True)

    # Add the direction (changes-for)
    add_changes_for(parser)

    # Add show reverse option
    add_reverse(parser)

    # Add regexp
    add_regexp(parser)

    # Add ssl options
    add_ssl_options(parser)

    # Now we process the rest of the arguments.
    opt, args = parser.parse_args()

    # Fail if no db arguments or --all option.
    if len(args) == 0 and not opt.all:
        parser.error(PARSE_ERR_DB_MISSING_CMP)

    # Check security settings
    check_password_security(opt, args, "# ")

    # Warn if quiet and verbosity are both specified
    check_verbosity(opt)

    # Fail if the --all option and database arguments are both specified.
    check_all(parser, opt, args, "databases")

    # Process --exclude values to remove unnecessary quotes (when used) in
    # order to avoid further matching issues.
    exclude_list = None
    if opt.exclude:
        if not opt.all:
            if not opt.quiet:
                print(WARN_OPT_ONLY_USED_WITH.format(opt='--exclude',
                                                     used_with='the --all'
                                                               ' option'))
        else:
            # Remove unnecessary outer quotes.
            exclude_list = [pattern.strip("'\"") for pattern in opt.exclude]

    # The --regexp option requires --exclude.
    if opt.use_regexp and not opt.exclude:
        if not opt.quiet:
            print(WARN_OPT_ONLY_USED_WITH.format(opt='--regexp',
                                                 used_with='the --exclude'
                                                           ' option'))

    # Check for regexp symbols
    check_exclude_pattern(exclude_list, opt.use_regexp)

    db_idxes_l = None

    # Set options for database operations.
    options = {
        "quiet": opt.quiet,
        "verbosity": opt.verbosity,
        "difftype": opt.difftype,
        "run_all_tests": opt.run_all_tests,
        "width": opt.width,
        "no_checksum_table": opt.no_checksum_table,
        "no_object_check": opt.no_object_check,
        "no_diff": opt.no_diff,
        "no_row_count": opt.no_row_count,
        "no_data": opt.no_data,
        "format": opt.format,
        "toggle_binlog": opt.toggle_binlog,
        "changes-for": opt.changes_for,
        "reverse": opt.reverse,
        "span_key_size": opt.span_key_size,
        "skip_opt_autoinc": opt.skip_opt_autoinc,
        "skip_partitioning": opt.skip_partitioning,
        "skip_table_opts": opt.skip_tbl_opts,
        "charset": opt.charset,
        "use_indexes": db_idxes_l,
        "compact": opt.compact,
        "all": opt.all,
        "use_regexp": opt.use_regexp,
        "exclude_patterns": exclude_list,
    }

    # Add ssl options to options instead of connection.
    options.update(get_ssl_dict(opt))

    # Check server options.
    if not opt.server1:
        parser.error(PARSE_ERR_OPTS_REQ.format(opt='--server1'))
    if opt.all and not opt.server2:
        parser.error(PARSE_ERR_OPTS_REQ.format(opt='--server2'))

    # Parse server connection values
    server1_values = None
    server2_values = None
    try:
        server1_values = parse_connection(opt.server1, None, options)
    except FormatError:
        _, err, _ = sys.exc_info()
        parser.error("Server1 connection values invalid: %s." % err)
    except UtilError:
        _, err, _ = sys.exc_info()
        parser.error("Server1 connection values invalid: %s." % err.errmsg)

    if opt.server2:
        try:
            server2_values = parse_connection(opt.server2, None, options)
        except FormatError:
            _, err, _ = sys.exc_info()
            parser.error("Server2 connection values invalid: %s." % err)
        except UtilError:
            _, err, _ = sys.exc_info()
            parser.error("Server2 connection values invalid: %s." % err.errmsg)
    else:
        server2_values = None

    # Get the sql_mode set on source and destination server
    conn_opts = {
        'quiet': True,
        'version': "5.1.30",
    }
    try:
        servers = connect_servers(server1_values, server2_values, conn_opts)
        server1_sql_mode = servers[0].select_variable("SQL_MODE")
        if servers[1] is not None:
            server2_sql_mode = servers[1].select_variable("SQL_MODE")
        else:
            server2_sql_mode = ''
    except UtilError:
        server1_sql_mode = ''
        server2_sql_mode = ''

    ukey_regex_server1 = REGEXP_OBJ_NAME
    if "ANSI_QUOTES" in server1_sql_mode:
        ukey_regex_server1 = REGEXP_OBJ_NAME_AQ

    ukey_regex_server2 = REGEXP_OBJ_NAME
    if "ANSI_QUOTES" in server2_sql_mode:
        ukey_regex_server2 = REGEXP_OBJ_NAME_AQ

    # check unique keys
    ukey_regexp = re.compile(r'(?:(?:;){{0,1}}{0}\.{1})'
                             ''.format(ukey_regex_server1, ukey_regex_server2))

    # Split the table names considering backtick quotes
    if opt.use_indexes:
        grp = ukey_regexp.findall(opt.use_indexes)
        if not grp:
            parser.error("Can't parse the specified --use-indexes argument {0}"
                         "".format(opt.use_indexes))
        db_idxes_l = []
        for table, index in grp:
            table_uc = (table if is_quoted_with_backticks(table,
                                                          server1_sql_mode)
                        else quote_with_backticks(table, server1_sql_mode))
            index_uc = (index if is_quoted_with_backticks(index,
                                                          server1_sql_mode)
                        else quote_with_backticks(index, server1_sql_mode))
            db_idxes_l.append((table_uc, index_uc))
        options["use_indexes"] = db_idxes_l

    # Check --span-key-size value.
    if opt.span_key_size is not None:
        if opt.span_key_size < DEFAULT_SPAN_KEY_SIZE:
            parser.error(
                PARSE_ERR_SPAN_KEY_SIZE_TOO_LOW.format(
                    s_value=opt.span_key_size, default=DEFAULT_SPAN_KEY_SIZE))
        if opt.span_key_size > MAX_SPAN_KEY_SIZE:
            parser.error(
                PARSE_ERR_SPAN_KEY_SIZE_TOO_HIGH.format(
                    s_value=opt.span_key_size, max=MAX_SPAN_KEY_SIZE))
        if opt.span_key_size % 2 != 0:
            if not opt.quiet:
                print("# WARNING: The value for the --span-key-size option"
                      " must be an even number. The value {0} will be used "
                      "instead.".format(opt.span_key_size - 1))

    # Operations to perform:
    # 1) databases exist
    # 2) check object counts
    # 3) check object differences
    # 4) check row counts among the tables
    # 5) check table data consistency

    res = True
    check_failed = False

    if opt.all:
        # Compare all databases.
        try:
            res = compare_all_databases(server1_values, server2_values,
                                        exclude_list, options)
        except UtilError:
            _, e, _ = sys.exc_info()
            print("ERROR: %s" % e.errmsg)
            sys.exit(1)

        if res is None:
            check_failed = None
        elif not res:
            check_failed = True
    else:
        # Compare specified databases.
        arg_regexp = re.compile(r'{0}(?:(?:\:){0})?'.format(REGEXP_OBJ_NAME))
        for db in args:
            # Split the database names considering backtick quotes
            grp = arg_regexp.match(db)
            if not grp:
                parser.error(PARSE_ERR_DB_PAIR.format(db_pair=db,
                                                      db1_label='db1',
                                                      db2_label='db2'))
            parts = grp.groups()
            matched_size = len(parts[0])
            if not parts[1]:
                parts = (parts[0], parts[0])
            else:
                # add 1 for the separator ':'
                matched_size += 1
                matched_size += len(parts[1])
            # Verify if the size of the databases matched by the REGEX is equal
            # to the initial specified string. In general, this identifies the
            # missing use of backticks.
            if matched_size != len(db):
                parser.error(PARSE_ERR_DB_PAIR_EXT.format(db_pair=db,
                                                          db1_label='db1',
                                                          db2_label='db2',
                                                          db1_value=parts[0],
                                                          db2_value=parts[1]))

            # Remove backtick quotes (handled later)
            db1 = remove_backtick_quoting(parts[0], server1_sql_mode) \
                if is_quoted_with_backticks(parts[0], server1_sql_mode) \
                else parts[0]
            db2 = remove_backtick_quoting(parts[1], server1_sql_mode) \
                if is_quoted_with_backticks(parts[1], server1_sql_mode) \
                else parts[1]

            try:
                res = database_compare(server1_values, server2_values,
                                       db1, db2, options)
                if not opt.quiet:
                    print
            except UtilError:
                _, e, _ = sys.exc_info()
                print("ERROR: %s" % e.errmsg)
                sys.exit(1)

            if not res:
                check_failed = True

            if check_failed and not opt.run_all_tests:
                break

    if not opt.quiet:
        print
        if check_failed is None:
            print("# No databases to compare.")
        elif check_failed:
            print("# Database consistency check failed.")
        else:
            sys.stdout.write("# Databases are consistent")
            if (opt.no_object_check or opt.no_diff or
                    opt.no_row_count or opt.no_data or opt.skip_tbl_opts):
                sys.stdout.write(" given skip options specified")
            print(".")
        print("#\n# ...done")

    if check_failed:
        sys.exit(1)

    sys.exit()
